File: 684-redundant-connection/684-redundant-connection.java
class Solution {
    int[] parent, rank;
    public int[] findRedundantConnection(int[][] edges) {
        parent = new int[edges.length + 1];
        rank = new int[edges.length + 1];
        Arrays.fill(rank, 1);
        Arrays.setAll(parent, i -> i);

        for(int i = 0; i < edges.length; i++){
            int n1 = edges[i][0], n2 = edges[i][1];
            if(union(n1, n2))   continue;
            else    return edges[i];
        }
        return new int[2];
    }
    
    public boolean union(int n1, int n2){
         int p1 = find(n1), p2 = find(n2);
         if(p1 == p2)   return false;
         else if(rank[p1] > rank[p2]){
             parent[p2] = p1;
             rank[p1] += rank[p2];
         }
         else{
             parent[p1] = p2;
             rank[p2] += rank[p1];
         }
         return true;
     }
    
    public int find(int n){
        int p = parent[n];
        
        while(p != parent[p]){
            parent[p] = parent[parent[p]];
            p = parent[p];
        }
        
        return p;
    }
}

File: 684-redundant-connection/684-redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        parent = [i for i in range(len(edges) + 1)]
        rank = [1] * (len(edges) + 1)
        
        def find(n):
            p = parent[n]
            while p != parent[p]:
                parent[p] = parent[parent[p]]
                p = parent[p]    
            return p
        
        # Returning False if they are already merged
        def union(n1, n2):
            p1, p2 = find(n1), find(n2)
            if p1 == p2:
                return False
            
            if rank[p1] > rank[p2]:
                parent[p2] = p1
                rank[p1] += rank[p2]
            else:
                parent[p1] = p2
                rank[p2] += rank[p1]
                
            return True
        
        for n1, n2 in edges:
            if union(n1, n2):
                continue
            else:
                return [n1, n2]